hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/access/TxPutFromLoadInterceptor.java

```diff
@@ -0,0 +1,126 @@
+import org.infinispan.commands.write.InvalidateCommand;
+import org.infinispan.commands.write.WriteCommand;
+import org.infinispan.container.DataContainer;
+import org.infinispan.context.impl.TxInvocationContext;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.interceptors.base.BaseRpcInterceptor;
+import org.infinispan.remoting.inboundhandler.DeliverOrder;
+import org.infinispan.remoting.rpc.RpcManager;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.Set;
+
+/**
+ * Intercepts transactions in Infinispan, calling {@link PutFromLoadValidator#beginInvalidatingKey(Object, Object)}
+ * before locks are acquired (and the entry is invalidated) and sends {@link EndInvalidationCommand} to release
+ * invalidation throught {@link PutFromLoadValidator#endInvalidatingKey(Object, Object)} after the transaction
+ * is committed.
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+class TxPutFromLoadInterceptor extends BaseRpcInterceptor {
+	private final static Log log = LogFactory.getLog(TxPutFromLoadInterceptor.class);
+	private PutFromLoadValidator putFromLoadValidator;
+	private final String cacheName;
+	private RpcManager rpcManager;
+	private CacheCommandInitializer cacheCommandInitializer;
+	private DataContainer dataContainer;
+
+	public TxPutFromLoadInterceptor(PutFromLoadValidator putFromLoadValidator, String cacheName) {
+		this.putFromLoadValidator = putFromLoadValidator;
+		this.cacheName = cacheName;
+	}
+
+	@Inject
+	public void injectDependencies(RpcManager rpcManager, CacheCommandInitializer cacheCommandInitializer, DataContainer dataContainer) {
+		this.rpcManager = rpcManager;
+		this.cacheCommandInitializer = cacheCommandInitializer;
+		this.dataContainer = dataContainer;
+	}
+
+	// We need to intercept PrepareCommand, not InvalidateCommand since the interception takes
+	// place before EntryWrappingInterceptor and the PrepareCommand is multiplexed into InvalidateCommands
+	// as part of EntryWrappingInterceptor
+	@Override
+	public Object visitPrepareCommand(TxInvocationContext ctx, PrepareCommand command) throws Throwable {
+		if (ctx.isOriginLocal()) {
+			// We can't wait to commit phase to remove the entry locally (invalidations are processed in 1pc
+			// on remote nodes, so only local case matters here). The problem is that while the entry is locked
+			// reads still can take place and we can read outdated collection after reading updated entity
+			// owning this collection from DB; when this happens, the version lock on entity cannot protect
+			// us against concurrent modification of the collection. Therefore, we need to remove the entry
+			// here (even without lock!) and let possible update happen in commit phase.
+			for (WriteCommand wc : command.getModifications()) {
+				if (wc instanceof InvalidateCommand) {
+					// InvalidateCommand does not correctly implement getAffectedKeys()
```

Comment:
Should this be fixed in Infinispan? If so, can you create a JIRA for it if you've not already done so?